### ISSUE TITLE
Error running scenario_configs/poisoning_gtsrb_dirty_label.json with batch_size of 1

### ISSUE THREAD

**Issue by lcadalzo** [twosixlabs/armory#1537]
The error only seems to occur when batch_size is set to 1:

```
Traceback (most recent call last):

  File "/opt/conda/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
           │         │     └ {'__name__': '__main__', '__doc__': '\nMain script for running scenarios. Users will run a scenario by calling:\n    armory r...
           │         └ <code object <module> at 0x7f1cde9f8870, file "/workspace/armory/scenarios/main.py", line 1>
           └ <function _run_code at 0x7f1e190f4d30>
  File "/opt/conda/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
         │     └ {'__name__': '__main__', '__doc__': '\nMain script for running scenarios. Users will run a scenario by calling:\n    armory r...
         └ <code object <module> at 0x7f1cde9f8870, file "/workspace/armory/scenarios/main.py", line 1>

  File "/workspace/armory/scenarios/main.py", line 228, in <module>
    run_config(
    └ <function run_config at 0x7f1cdd850280>

  File "/workspace/armory/scenarios/main.py", line 144, in run_config
    scenario.evaluate()
    │        └ <function Scenario.evaluate at 0x7f1cdd50bdc0>
    └ <armory.scenarios.poison.Poison object at 0x7f1e187cea00>

> File "/workspace/armory/scenarios/scenario.py", line 390, in evaluate
    self._evaluate()
    │    └ <function Scenario._evaluate at 0x7f1cdd50bd30>
    └ <armory.scenarios.poison.Poison object at 0x7f1e187cea00>

  File "/workspace/armory/scenarios/scenario.py", line 379, in _evaluate
    self.load()
    │    └ <function Poison.load at 0x7f1cdd50fc10>
    └ <armory.scenarios.poison.Poison object at 0x7f1e187cea00>

  File "/workspace/armory/scenarios/poison.py", line 460, in load
    self.fit()
    │    └ <function Poison.fit at 0x7f1cdd50f8b0>
    └ <armory.scenarios.poison.Poison object at 0x7f1e187cea00>

  File "/workspace/armory/scenarios/poison.py", line 313, in fit
    self.model.fit(
    │    │     └ <function InputFilter.__init__.<locals>.make_replacement.<locals>.replacement_function at 0x7f1ce57fc4c0>
    │    └ art.estimators.classification.pytorch.PyTorchClassifier(model=ModelWrapper(
    │        (_model): Sequential(
    │          (0): Permute()
    │          (1...
    └ <armory.scenarios.poison.Poison object at 0x7f1e187cea00>

  File "/opt/conda/lib/python3.8/site-packages/art/estimators/classification/classifier.py", line 71, in replacement_function
    return fdict[func_name](self, *args, **kwargs)
           │     │          │      │       └ {'batch_size': 1, 'nb_epochs': 1, 'verbose': False, 'shuffle': True}
           │     │          │      └ (array([[[[0.1882353 , 0.18431373, 0.17254902],
           │     │          │                 [0.15294118, 0.14901961, 0.13333334],
           │     │          │                 [0.09019608, 0.090196...
           │     │          └ art.estimators.classification.pytorch.PyTorchClassifier(model=ModelWrapper(
           │     │              (_model): Sequential(
           │     │                (0): Permute()
           │     │                (1...
           │     └ 'fit'
           └ {'__module__': 'art.estimators.classification.pytorch', '__qualname__': 'PyTorchClassifier', '__doc__': '\n    This class imp...
  File "/opt/conda/lib/python3.8/site-packages/art/estimators/classification/pytorch.py", line 412, in fit
    model_outputs = self._model(i_batch)
                    │    │      └ tensor([[[[0.0353, 0.3059, 0.2235],
                    │    │                  [0.0549, 0.3412, 0.2980],
                    │    │                  [0.0510, 0.3843, 0.3490],
                    │    │                  ...,
                    │    │          ...
                    │    └ ModelWrapper(
                    │        (_model): Sequential(
                    │          (0): Permute()
                    │          (1): Conv2d(3, 1, kernel_size=(1, 1), stride=(1, 1))
                    │          (2): Ba...
                    └ art.estimators.classification.pytorch.PyTorchClassifier(model=ModelWrapper(
                        (_model): Sequential(
                          (0): Permute()
                          (1...
  File "/opt/conda/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
           │             │        └ {}
           │             └ (tensor([[[[0.0353, 0.3059, 0.2235],
           │                         [0.0549, 0.3412, 0.2980],
           │                         [0.0510, 0.3843, 0.3490],
           │                         ...,
           │                ...
           └ <bound method PyTorchClassifier._make_model_wrapper.<locals>.ModelWrapper.forward of ModelWrapper(
               (_model): Sequential(
               ...
  File "/opt/conda/lib/python3.8/site-packages/art/estimators/classification/pytorch.py", line 1115, in forward
    x = module_(x)
        │       └ tensor([[0.6033, 1.1313, 0.3681, 0.0000, 0.7217, 0.0000, 0.0000, 0.0000, 0.5263,
        │                  0.4399, 0.0000, 0.0000, 0.0000, 0.2...
        └ BatchNorm1d(300, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  File "/opt/conda/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
           │             │        └ {}
           │             └ (tensor([[0.6033, 1.1313, 0.3681, 0.0000, 0.7217, 0.0000, 0.0000, 0.0000, 0.5263,
           │                        0.4399, 0.0000, 0.0000, 0.0000, 0....
           └ <bound method _BatchNorm.forward of BatchNorm1d(300, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)>
  File "/opt/conda/lib/python3.8/site-packages/torch/nn/modules/batchnorm.py", line 168, in forward
    return F.batch_norm(
           │ └ <function batch_norm at 0x7f1d2d2c0550>
           └ <module 'torch.nn.functional' from '/opt/conda/lib/python3.8/site-packages/torch/nn/functional.py'>
  File "/opt/conda/lib/python3.8/site-packages/torch/nn/functional.py", line 2280, in batch_norm
    _verify_batch_size(input.size())
    │                  │     └ <method 'size' of 'torch._C._TensorBase' objects>
    │                  └ tensor([[0.6033, 1.1313, 0.3681, 0.0000, 0.7217, 0.0000, 0.0000, 0.0000, 0.5263,
    │                             0.4399, 0.0000, 0.0000, 0.0000, 0.2...
    └ <function _verify_batch_size at 0x7f1d2d2c04c0>
  File "/opt/conda/lib/python3.8/site-packages/torch/nn/functional.py", line 2248, in _verify_batch_size
    raise ValueError("Expected more than 1 value per channel when training, got input size {}".format(size))
                                                                                                      └ torch.Size([1, 300])

ValueError: Expected more than 1 value per channel when training, got input size torch.Size([1, 300])

```

**Comment by lcadalzo** [comment]
The error occurs in [this](https://github.com/pytorch/pytorch/blob/v1.10.2/torch/nn/functional.py#L2234-L2248) function because the `size_prods` variable is equal to batch_size, and when `size_prods` is 1, this function raises a ValueError. 

This function is called by `torch.nn.functional.batch_norm()` during training, which can be seen [here](https://github.com/pytorch/pytorch/blob/v1.10.2/torch/nn/functional.py#L2280), so it appears the error is essentially that torch is complaining about doing batch norm with only 1 element

**Comment by lcadalzo** [comment]
I noticed that the issue does not occur for `scenario_configs/eval5/poisoning/gtsrb_witches_brew.json`, which uses the same model. The reason is that at [this](https://github.com/pytorch/pytorch/blob/v1.10.2/torch/nn/modules/batchnorm.py#L158) line, `self.traning` is `False` for some reason (it's `True` for the config in the PR title). Thus, because the `training` kwarg is False, the function `_verify_batch_size()` is never called (see [this](https://github.com/pytorch/pytorch/blob/v1.10.2/torch/nn/functional.py#L2279) conditional) and so the error doesn't occur. Still unclear to me why `self.training` is `True` in one case but `False` in the other, though

**Comment by lcadalzo** [comment]
Turns out the error does in fact occur for `scenario_configs/eval5/poisoning/gtsrb_witches_brew.json` (only after a lengthy `poison_dataset()` method, which I originally ctrl-c'ed out of because I was expecting that if the error was there it would occur quickly like it did for `poisoning_gtsrb_dirty_label`). It also occurs for other configs that use the gtsrb baseline model

A much clearer error than the one above would simply instruct the user to use batch_size > 1. I think I'd rather put that constraint in the model code rather than scenario code, since it need not apply to every model.

**Comment by lcadalzo** [comment]
I've noticed the error occurs for batch sizes larger than 1 (up to 8 is what I've noticed):
```
ValueError: Expected more than 1 value per channel when training, got input size torch.Size([1, 300])
root@c2dbd5883c4a:/workspace# cat /armory/tmp/2022-05-27T175841.540582/interactive-config.json | grep batch_size
        "batch_size": 8,

```

The smallest batch_size I tried that did not yield the error (at least within the ~5min or so I gave it to, I didn't run the config entirely to completion) was 12

**Comment by lcadalzo** [comment]
I've noticed the error occurs for batch sizes larger than 1 (up to 8 is what I've noticed):
```
ValueError: Expected more than 1 value per channel when training, got input size torch.Size([1, 300])
root@c2dbd5883c4a:/workspace# cat /armory/tmp/2022-05-27T175841.540582/interactive-config.json | grep batch_size
        "batch_size": 8,

```

The smallest batch_size I tried that did not yield the error (at least within the ~5min or so I gave it to, I didn't run the config entirely to completion) was 12

**Comment by lcadalzo** [comment]
I've noticed the error occurs for batch sizes larger than 1 (up to 8 is what I've noticed):
```
ValueError: Expected more than 1 value per channel when training, got input size torch.Size([1, 300])
root@c2dbd5883c4a:/workspace# cat /armory/tmp/2022-05-27T175841.540582/interactive-config.json | grep batch_size
        "batch_size": 8,

```

The smallest batch_size I tried that did not yield the error (at least within the ~5min or so I gave it to, I didn't run the config entirely to completion) was 12

**Comment by lcadalzo** [comment]
The error occurs even with batch_size of 8 because the very last batch in the epoch is of size 1. The dataset has 39209 elements and `39209 % 8 = 1` . It also occured with batch_size of 2, which makes sense since `39209 % 2 = 1` as well. This doesn't appear to be an Armory error

**Comment by lcadalzo** [comment]
Added https://github.com/Trusted-AI/adversarial-robustness-toolbox/issues/1723 in ART

**Comment by mwartell** [comment]
https://github.com/Trusted-AI/adversarial-robustness-toolbox/issues/1723 was deferred from ART 1.11

**Comment by mwartell** [comment]
https://github.com/Trusted-AI/adversarial-robustness-toolbox/issues/1723 was deferred from ART 1.11

**Comment by davidslater** [comment]
How about we simply enforce `batch_size > 1` for this? (And give a small note about the ART issue.)

**Comment by lcadalzo** [comment]
haven't re-tested this since my earlier comments, but the error was occurring for batch size greater than 1 too (e.g. 2, 8, etc.) as mentioned in the comments above. Any batch size where `dataset size % batch_size = 1`

**Comment by lcadalzo** [comment]
Yea I think this feels like something that can hopefully be addressed in ART
